opcua/client/client.py

```diff
@@ -38,27 +38,39 @@ def __init__(self, client, timeout):
         self.client = client
         self._dostop = False
         self._cond = Condition()
+        self._interruption_callbacks = set()
 
     def run(self):
         self.logger.debug("starting keepalive thread with period of %s milliseconds", self.timeout)
-        server_state = self.client.get_node(ua.FourByteNodeId(ua.ObjectIds.Server_ServerStatus_State))
-        while not self._dostop:
-            with self._cond:
-                self._cond.wait(self.timeout / 1000)
-            if self._dostop:
-                break
-            self.logger.debug("renewing channel")
-            self.client.open_secure_channel(renew=True)
-            val = server_state.get_value()
-            self.logger.debug("server state is: %s ", val)
-        self.logger.debug("keepalive thread has stopped")
+        try:
+            server_state = self.client.get_node(ua.FourByteNodeId(ua.ObjectIds.Server_ServerStatus_State))
+            while not self._dostop:
+                with self._cond:
+                    self._cond.wait(self.timeout / 1000)
+                if self._dostop:
+                    break
+                self.logger.debug("renewing channel")
+                self.client.open_secure_channel(renew=True)
+                val = server_state.get_value()
+                self.logger.debug("server state is: %s ", val)
+        except Exception as e:
+            self._connection_interrupted(e)
+        finally:
+            self.logger.debug("keepalive thread has stopped")
 
     def stop(self):
         self.logger.debug("stoping keepalive thread")
         self._dostop = True
         with self._cond:
             self._cond.notify_all()
 
+    def register_interruption_callback(self, callback):
+        self._interruption_callbacks.add(callback)
+
+    def _connection_interrupted(self, exception):
+        # copy the set for thread safety
+        for callback in self._interruption_callbacks.copy():
```

Comment:
No, I actually meant using Lock object for registering and calling `self._interruption_callbacks` instead of `self._interruption_callbacks.copy()`.